Improvements to overloading resolution code.

paulp opened the pull request:
Review by @retronym.

paulp committed:
Replaced some <code> comments.

With the far more readable not-html comments of modern times.

paulp committed:
Worked over inferMethodAlternative.

We shouldn't let these core methods become so complicated. They are
way harder to follow than is necessary. It is why nobody ever fixes
the bugs in them. This is only the beginning really. So many things
in the compiler would border on trivial to fix if one didn't have to
navigate so much cruft and indirection.

Here are some methods which no longer exist:

 - hasExactlyNumParams. A dubious name for a method
   containing the expression "len <= n + 1".

 - resolveOverloadedMethod. A method which returns a list
   of symbols can't be resolving all that much.

 - isUnitForVarArgs. Take a guess as to what that method does.
   Ha ha, you were not even close.

Still on my hit list:

 - "very similar logic to doTypedApply in typechecker"
   I find a good rule of thumb is never to write a comment
   which paraphrases to "this is very similar to that."

This entire patch is the fault of a. moors for trying to
cherry-pick a comment of mine from github into trunk. This
patch hopefully makes the comment unnecessary.

paulp commented:
I forgot to mention: this increases the uniformity of followApply, but not all the way.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/699/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1408/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1408/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/699/

paulp commented:
I borked a test with a last minute change; don't let that lone failure distract you, it's all quite sound.

retronym commented:
Looks like a big improvement, give me a day or two to review it in detail.

hubertp commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -49,6 +49,24 @@ trait Infer extends Checkable {
+
+    alternatives sortWith improves match {
+      case best :: rest if rest.nonEmpty => best :: rest.filterNot(alt => improves(best, alt))
+      case bests                         => bests
```
Wouldn't best :: Nil be more readable?

paulp replied:
I'm not sure which case you mean, but neither of those cases simplifies to best :: Nil.  If you mean the second case, it also includes Nil (which does come through in practice.)

hubertp replied:
True, got confused by the rest.nonEmpty. Looks correct.

hubertp commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -672,6 +679,35 @@ trait Infer extends Checkable {
+        // with the same number of parameters.  See SI-5859 for an example of what
+        // would fail were this not enforced before we arrived at isApplicable.
+        if (varargsStar)
+          varargsTarget && (paramsCount == argsCount)
```
paramsCount == argsCount ==> simpleMatch

paulp replied:
Good point!

hubertp commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -1489,52 +1539,29 @@ trait Infer extends Checkable {
+        bests match {
+          case best :: Nil                              => tree setSymbol best setType (pre memberType best)
+          case best :: competing :: _ if alts0.nonEmpty => if (!pt.isErroneous) AmbiguousExprAlternativeError(tree, pre, best, competing, pt, isSecondTry)
+          case _                                        => if (bests.isEmpty || alts0.isEmpty) NoBestExprAlternativeError(tree, pt, isSecondTry) // todo: missing test case
```
the //todo is actually a bit misleading. we are missing test only for bests.isEmpty.

paulp replied:
My fault; I knew that was a lossy comment move when I did it.

paulp committed:
Merge branch 'master' into infer-method-alternative

paulp committed:
Incorporated pull request feedback.

And fixed the test I broke at the last minute. Reworked
tupling logic to make it harder to break.  Expanded test
coverage.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/703/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1412/

hubertp commented:
I wasn't assigned as a reviewer of this one but I spent my share of time in those methods and this is definitely an improvement. so lgtm.

paulp commented:
Thanks.  Clearly I should be shopping at your review store more often.

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1412/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/703/

paulp commented:
@retronym, I am still keen for any input you would like to offer, but I am even more keen not to let my pull requests stack up when I have this lgtm and a pair of jenkins successes burning a hole in my pocket.  In other words, I'm merging.
